src/zkevm_specs/evm/instruction.py

```diff
@@ -368,6 +372,43 @@ def range_check(self, value: Expression, n_bytes: int) -> bytes:
         except OverflowError:
             raise ConstraintUnsatFailure(f"Value {value} has too many bytes to fit {n_bytes} bytes")
 
+    # Return a tuple of `abs(x)` and `x_is_neg`. For a special case when
+    # `x = -(1 << 255)`, this function returns the same value of `-(1 << 255)`,
+    # since it is signed overflow.
+    def abs_word(self, x: RLC) -> Tuple[RLC, FQ]:
+        is_neg = self.word_is_neg(x)
+
+        # Generate the witness `x_abs`.
+        x_abs = x if is_neg == 0 else self.rlc_encode((1 << 256) - x.int_value, 32)
+
+        x_abs_lo, x_abs_hi = self.word_to_lo_hi(x_abs)
+        x_lo, x_hi = self.word_to_lo_hi(x)
+
+        # Constrain `x_abs_lo == x_lo` and `x_abs_hi == x_hi` if non negative.
+        self.constrain_zero((x_abs_lo - x_lo) * (1 - is_neg))
+        self.constrain_zero((x_abs_hi - x_hi) * (1 - is_neg))
+
+        # When `is_neg`, contrain `x + x_abs == 1 << 256`. Even if
+        # `x = -(1 << 255)` that is signed overflow, and
+        # `abs(-(1 << 255) = -(1 << 255)`.
+        carry_lo, sum_lo = divmod(x_lo.n + x_abs_lo.n, 1 << 128)
+        carry_hi, sum_hi = divmod(x_hi.n + x_abs_hi.n + carry_lo, 1 << 128)
+
+        # Contrain `sum([x_lo, x_abs_lo]) == sum_lo + carry_lo * 2^128`.
+        self.constrain_zero(FQ(sum_lo) + FQ(carry_lo) * FQ(1 << 128) - self.sum([x_lo, x_abs_lo]))
```

Comment:
I don't think we need to check `carry_hi` and `carry_lo` to be binary because the bytes range check for `x_lo` and `x_abs_lo` (same as `x_hi` and `x_abs_hi`) constrains these two values are within [0, 2^128). So `x_lo + x_abs_lo` must be in the range [0, 2^129), which entails that `carry_lo` can only be 0 or 1.